### PR TITLE
Support tied parameters

### PR DESCRIPTION
**Pull request by ffuuugor**
## Solution

As reported in #413, some models (including huggingface GPT-2) use tied parameters - i.e. sharing the same parameter between modules by simply using the same Parameter object. 
Opacus gets confused - it treats the second set of grad_samples as a new batch and accumulates them by stacking, which is obviously wrong. Conceptually this is not very different from RNNs, and the gradients should be simply added up.

Natural way to support this is to tweak the condition on which we distinguish between recurrent calls and new batch. Currently we look at module-level activations - if we've run out of the activations, it means that the recurrent calls are over and the next grad sample will be from a new batch. This obviously doesn't work for shared parameters - as the condition is module-level, but the parameter belong to multiple modules.

What we can do instead is to have parameter-level counters (`p._forward_counter`). This counter would increase on forward pass and decrease on backward. When it gets to zero, in would indicate that the latest backward pass is fully unrolled and we should expect new forward pass with the new data. If I understand correctly, we do not and should not allow multiple forwards or backwards in a row.

## Tests

I've added a new unit test for this and also checked Colab from the bug report - all seem to be working.

### CODE REVIEW

File: opacus/grad_sample/grad_sample_module.py [hunk 1]
```diff
@@ -296,6 +303,9 @@ def capture_activations_hook(
             module.activations = []
         module.activations.append(forward_input[0].detach())  # pyre-ignore
 
+        for p in module.parameters():
```
Comment (karthikprasad) [inline opacus/grad_sample/grad_sample_module.py:306]: Do you think it will help to add a comment above this line mentioning this is for tied parameters? A link to the PR itself is even better. Your PR summary is stellar!

File: opacus/tests/privacy_engine_test.py [hunk 2]
```diff
@@ -262,7 +262,9 @@ def test_compare_to_vanilla(self):
         for do_noise in (True, False):
             for do_clip in (True, False):
                 for use_closure in (True, False):
-                    with self.subTest(do_noise=do_noise, do_clip=do_clip):
+                    with self.subTest(
+                        do_noise=do_noise, do_clip=do_clip, use_closure=use_closure
```
Comment (karthikprasad) [inline opacus/tests/privacy_engine_test.py:266]: Great catch!

File: opacus/grad_sample/grad_sample_module.py [hunk 3]
```diff
@@ -341,15 +351,19 @@ def capture_backprops_hook(
...
         for param, gs in grad_samples.items():
-            create_or_accumulate_grad_sample(param=param, grad_sample=gs, layer=module)
+            create_or_accumulate_grad_sample(
+                param=param, grad_sample=gs, max_batch_len=module.max_batch_len
+            )
+
+        for p in module.parameters():
+            p._forward_counter -= 1
+            if p._forward_counter == 0:
+                promote_current_grad_sample(p)
```
Comment (karthikprasad) [inline opacus/grad_sample/grad_sample_module.py:361]: nit: I suppose this can still be inside the `if len(module.activations) == 0` block?

**Review by karthikprasad — APPROVED** [review]
A1 code changes! ( a couple of nits)
A0 summary! <3

Thanks!

File: opacus/grad_sample/grad_sample_module.py [hunk 4]
```diff
@@ -125,6 +124,7 @@ def __init__(
 
         for p in self.parameters():
```
Comment (karthikprasad) [inline opacus/grad_sample/grad_sample_module.py:125]: <Comment unrelated to this specific PR/issue.>
This had not caught my eye before. Is there any reason why we attach `.grad_sample` to all parameters instead of just the trainable ones?

File: opacus/grad_sample/grad_sample_module.py (hunk 3 above)
Comment (ffuuugor) [reply opacus/grad_sample/grad_sample_module.py:361]: The check itself - yes, but the decrement - no.

File: opacus/grad_sample/grad_sample_module.py (hunk 4 above)
Comment (ffuuugor) [reply opacus/grad_sample/grad_sample_module.py:125]: Weeell, I don't think there's a reason. 
I always thought we've had a util method that does '[p for p in model.parameters() if p.requires_grad]', but apparently no, and we're consistently using just `model.parameters()` all over

**Comment by vdasu** [comment]
Is this functionality added to the latest version of Opacus? I noticed the PR is closed. I am facing the same issue when trying to train GPT-2
